src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/AlgorithmIdentifier.cs

```diff
@@ -27,9 +27,148 @@ public AlgorithmIdentifier(Oid oid, int keyLength)
             KeyLength = keyLength;
         }
 
+        internal AlgorithmIdentifier(CERT_PUBLIC_KEY_INFO keyInfo)
+        {
+            int keyLength = CertGetPublicKeyLength(MsgEncodingType.X509_ASN_ENCODING | MsgEncodingType.PKCS_7_ASN_ENCODING, ref keyInfo);
+            byte[] parameters = new byte[keyInfo.Algorithm.Parameters.cbData];
+            if (parameters.Length > 0)
+            {
+                Marshal.Copy(keyInfo.Algorithm.Parameters.pbData, parameters, 0, parameters.Length);
+            }
+
+            Reset(Oid.FromOidValue(Marshal.PtrToStringAnsi(keyInfo.Algorithm.pszObjId), OidGroup.PublicKeyAlgorithm), keyLength, parameters);
+        }
+
+        internal AlgorithmIdentifier(CRYPT_ALGORITHM_IDENTIFIER cryptAlgorithmIdentifer)
```

Comment:
Did you add these constructors because Mono uses them?  Otherwise they were fine as helper methods.

Reply:
Actually, these should go back to helper methods, because the Unix versions won't have crypt32 interop.

Reply:
At the moment the only need for the Mono is Parameters property :) But it also should be valuable for corefx so I'm trying to follow your comments. I've added these ctors similar to reference source implementation. I'll move it back to helper methods. Thanks for quick look!